client/client-lib/src/query.rs

```diff
@@ -0,0 +1,194 @@
+        return if let (peer, Ok(result)) = response {
+            self.results.extend(result.into_iter());
+            self.responses.insert(peer);
+
+            if self.responses.len() > max_evil {
+                QueryStep::Finished(Ok(self.results.drain().collect()))
+            } else {
+                QueryStep::Continue
+            }
+        } else {
+            // handle error case using the CurrentConsensus method
+            self.current.process(response, max_evil)
+        };
+    }
+}
+
+/// Returns when f+1 responses are equal, retrying on 404 errors
+pub struct Retry404<R> {
+    current: CurrentConsensus<R>,
+}
+
+impl<R> Default for Retry404<R> {
+    fn default() -> Self {
+        Self {
+            current: CurrentConsensus::default(),
+        }
+    }
+}
+
+impl<R: Hash + Eq + Clone> QueryStrategy<R> for Retry404<R> {
+    fn process(&mut self, response: FedResponse<R>, max_evil: usize) -> QueryStep<R> {
+        match response {
+            (peer, Err(JsonRpcError::Call(RpcCallError::Custom(e)))) if e.code() == 404 => {
+                QueryStep::Request(HashSet::from([peer]))
+            }
+            response => self.current.process(response, max_evil),
+        }
+    }
+}
+
+/// Returns when f+1 responses are equal, retrying after every 2f+1 responses
+pub struct EventuallyConsistent<R> {
+    responses: HashSet<PeerId>,
+    current: CurrentConsensus<R>,
+}
+
+impl<R> Default for EventuallyConsistent<R> {
+    fn default() -> Self {
+        Self {
+            responses: HashSet::new(),
+            current: CurrentConsensus::default(),
+        }
+    }
+}
+
+impl<R: Hash + Eq + Clone> QueryStrategy<R> for EventuallyConsistent<R> {
+    fn process(&mut self, response: FedResponse<R>, max_evil: usize) -> QueryStep<R> {
+        let honest_threshold = max_evil * 2 + 1;
+        self.responses.insert(response.0);
+
+        match self.current.process(response, max_evil) {
+            QueryStep::Continue if self.responses.len() == honest_threshold => {
+                let result = QueryStep::Request(self.responses.clone());
+                self.responses.clear();
+                result
+            }
+            result => result,
+        }
+    }
+}
+
+/// Returns when f+1 responses are equal
+pub struct CurrentConsensus<R> {
+    pub results: HashMap<R, HashSet<PeerId>>,
+    pub errors: HashMap<PeerId, JsonRpcError>,
+}
+
+impl<R> Default for CurrentConsensus<R> {
+    fn default() -> Self {
+        Self {
+            results: HashMap::new(),
+            errors: HashMap::new(),
+        }
+    }
+}
+
+impl<R: Hash + Eq + Clone> QueryStrategy<R> for CurrentConsensus<R> {
+    fn process(&mut self, response: FedResponse<R>, max_evil: usize) -> QueryStep<R> {
+        match response {
+            (peer, Ok(result)) => {
+                let peers = self
+                    .results
+                    .entry(result)
+                    .or_insert_with(|| HashSet::from([peer]));
```

Comment:
`.or_default().insert(peer)` ? The code is inserting kind of twice RN.

Do we need to check if the peer wasn't already there?

Reply:
Ah I'll can change to `HashSet::new()`